GVFS/GVFS.Common/Cleanup/PrefetchStep.cs

```diff
@@ -1,105 +1,147 @@
 
-        public static bool TryPrefetchCommitsAndTrees(
-            ITracer tracer,
-            GVFSEnlistment enlistment,
-            PhysicalFileSystem fileSystem,
-            GitObjects gitObjects,
-            out string error)
+        public PrefetchStep(GVFSContext context, GitObjects gitObjects, bool requireCacheLock)
+            : base(context, gitObjects, requireCacheLock)
         {
+        }
+
+        public override string TelemetryKey => "PrefetchStep";
+
+        public bool TryPrefetchCommitsAndTrees(out string error, GitProcess gitProcess = null)
+        {
+            if (gitProcess == null)
+            {
+                gitProcess = new GitProcess(this.Context.Enlistment);
+            }
+
             List<string> packIndexes;
             using (FileBasedLock prefetchLock = GVFSPlatform.Instance.CreateFileBasedLock(
-                fileSystem,
-                tracer,
-                Path.Combine(enlistment.GitPackRoot, PrefetchCommitsAndTreesLock)))
+                this.Context.FileSystem,
+                this.Context.Tracer,
+                Path.Combine(this.Context.Enlistment.GitPackRoot, PrefetchCommitsAndTreesLock)))
             {
-                WaitUntilLockIsAcquired(tracer, prefetchLock);
+                this.WaitUntilLockIsAcquired(this.Context.Tracer, prefetchLock);
                 long maxGoodTimeStamp;
 
-                gitObjects.DeleteStaleTempPrefetchPackAndIdxs();
+                this.GitObjects.DeleteStaleTempPrefetchPackAndIdxs();
+                this.GitObjects.DeleteTemporaryFiles();
 
-                if (!TryGetMaxGoodPrefetchTimestamp(tracer, enlistment, fileSystem, gitObjects, out maxGoodTimeStamp, out error))
+                if (!this.TryGetMaxGoodPrefetchTimestamp(out maxGoodTimeStamp, out error))
                 {
                     return false;
                 }
 
-                if (!gitObjects.TryDownloadPrefetchPacks(maxGoodTimeStamp, out packIndexes))
+                if (!this.GitObjects.TryDownloadPrefetchPacks(gitProcess, maxGoodTimeStamp, out packIndexes))
                 {
                     error = "Failed to download prefetch packs";
                     return false;
                 }
             }
 
-            if (packIndexes?.Count > 0)
+            this.TrySchedulePostFetchJob(packIndexes);
+
+            return true;
+        }
+
+        protected override void RunGitAction()
+        {
+            long last;
+            string error = null;
+
+            if (!this.TryGetMaxGoodPrefetchTimestamp(out last, out error))
             {
-                TrySchedulePostFetchJob(tracer, enlistment.NamedPipeName, packIndexes);
+                this.Context.Tracer.RelatedError(error);
+                return;
             }
 
-            return true;
+            DateTime lastDateTime = EpochConverter.FromUnixEpochSeconds(last);
+            DateTime now = DateTime.UtcNow;
+
+            if (now <= lastDateTime + this.timeBetweenPrefetches)

```

Comment:
Ah okay here is where it is making sure the prefetch step isn't running too often.  This almost seems like the schedulers job.  It should know when the step ran and finished so should it be checking time between and just not schedule the work?

Reply:
The scheduler is for running `Execute()` on some schedule.

The steps are responsible for checking if we should do anything during that `Execute()` step. For instance, the prefetch step can check the timestamps on the prefetch packs. The later steps can't use this, and they'll need their own mechanism.

Specifically for the larger maintenance steps in the future: we will need some custom data file in the object cache to describe how long it has been since it ran (say, about once a week for repack), but need to check shortly after mount because of users who restart their machines daily.